src/slicegpt/adapters/phi3_adapter.py

```diff
@@ -0,0 +1,276 @@
+
+        if output_attentions:
+            outputs += (self_attn_weights,)
+
+        if use_cache:
+            outputs += (present_key_value,)
+
+        return outputs
+
+
+class Phi3LayerAdapter(LayerAdapter):
+    def __init__(self, layer: Phi3DecoderLayer) -> None:
+        super().__init__()
+        self._layer: Phi3DecoderLayer = layer
+
+    @property
+    def layer(self) -> Module:
+        return self._layer
+
+    @property
+    def hidden_states_args_position(self) -> int:
+        return 0
+
+    @property
+    def hidden_states_output_position(self) -> int:
+        return 0
+
+    def get_first_layernorm(self) -> Module:
+        return self.layer.input_layernorm
+
+    def get_second_layernorm(self) -> Module:
+        return self.layer.post_attention_layernorm
+
+    def get_attention_inputs(self) -> list[Linear]:
+        return [self.layer.self_attn.qkv_proj]
+
+    def get_attention_output(self) -> Linear:
+        return self.layer.self_attn.o_proj
+
+    def get_mlp_inputs(self) -> list[Linear]:
+        return [self.layer.mlp.gate_up_proj]
+
+    def get_mlp_output(self) -> Linear:
+        return self.layer.mlp.down_proj
+
+
+class Phi3ModelAdapter(ModelAdapter):
+    def __init__(self, model: Phi3ForCausalLM) -> None:
+        super().__init__()
+        self._model: Phi3ForCausalLM = model
+
+    @property
+    def model(self) -> Module:
+        return self._model
+
+    @property
+    def config(self) -> PretrainedConfig:
+        return self._model.config
+
+    @property
+    def config_type(self) -> type:
+        return Phi3Config
+
+    @property
+    def parallel_blocks(self) -> bool:
+        return False
+
+    @property
+    def seqlen(self) -> int:
+        # if no sliding window, max_position_embeddings is same as original_max_position_embeddings
+        return self.config.max_position_embeddings
+
+    @property
+    def hidden_size(self) -> int:
+        return self.config.hidden_size
+
+    @property
+    def should_bake_mean_into_linear(self) -> bool:
+        return False
+
+    @property
+    def original_layer_type(self) -> type:
+        return Phi3DecoderLayer
+
+    @property
+    def original_layer_norm_type(self) -> type:
+        return Phi3RMSNorm
+
+    @property
+    def layer_adapter_type(self) -> type:
+        return Phi3LayerAdapter
+
+    @property
+    def compressed_layer_type(self) -> type:
+        return CompressedPhi3DecoderLayer
+
+    @property
+    def use_cache(self) -> bool:
+        return self.config.use_cache
+
+    @use_cache.setter
+    def use_cache(self, value: bool) -> None:
+        self.config.use_cache = value
+
+    def compute_output_logits(self, input_ids: Tensor) -> FloatTensor:
+        return self.model(input_ids=input_ids).logits
+
+    def convert_layer_to_compressed(self, layer: Module, layer_idx: int | None) -> Module:
+        compressed_layer = self.compressed_layer_type(cast(self.config_type, self.config), layer_idx).to(
```

Comment:
Is there a type-checker that finds casting to type variables useful? Pylance is certainly complaining for me. I suggest removing `cast` and adding a `# type: ignore`, if necessary.

Reply:
I think casting in Python is generally a noop. Only helpful for type-checkers and readability. It's a non-issue here. Removed.